### ISSUE TITLE
Broken UI under webkit2gtk 2.26

### ISSUE THREAD

**Issue by aidanholm** [luakit/luakit#804]
## I'm submitting a…
Bug Report

## Current Behavior
Broken UI under webkit2gtk 2.26: frozen webpages, favicons not showing

## Environment
Linux Distribution & Version: Arch

**Comment by alex3kov** [comment]
Two issues I''ve experienced so far:  
* When passing URLs to already running Luakit: it results in empty page and I have to reload it to display it
* Much higher CPU usage than on WebKit 2.24

**Comment by zappathustra** [comment]
Upgraded to webkit2gtk 2.26 on Arch: Luakit shows blank pages, no matter the URL. Not sure they are loaded, actually. Downgraded to 2.24 and it works fine.

**Comment by 7415963987456321** [comment]
Also have this issue, reloading seems to fix it.
Don't have any issues with favicons though.-
EDIT: Holy batman, didn't notice the high cpu, constantly running at ~80%.

**Comment by ghost** [comment]
> Upgraded to webkit2gtk 2.26 on Arch: Luakit shows blank pages, no matter the URL. Not sure they are loaded, actually. Downgraded to 2.24 and it works fine.

I have the same issue using webkit2gtk v2.26. I don't think they are loaded as verbose mode show that the URL is actually not requested until reload.

**Comment by aidanholm** [comment]
This is due to webkit2gtk 2.26 apparently no longer starting a web process for a webview until a page URL is set. Page loads are prevented by the adblock module until it's loaded its blocklists, which would happen on the web process. I'm not sure why reloading the page circumvents the load block from the adblock module.

The bigger issue is that every tab must now have its own independent web process. This means that opening a new tab must now block until the web process loads, including all lua web modules, causing a noticable delay.

My CPU usage is high, but not *that* high: I have CPU use around 10% from luakit on 2.26: seems to be caused by the awful hack used to detect document resizes. Still waiting for ResizeObserver to be enabled for a true fix. An interim measure would be switching to setTimeout-based polling instead of requestAnimationFrame-basde polling, since the latter forces constant redraws in the luakit process.

**Comment by gersonamoriz** [comment]
I also having a very high CPU peak with webkit2gtk 2.26, and same problem with blank pages, CPU is so high that I had to downgrade the pkg to 2.24, still working fine after downgrade.

> Two issues I''ve experienced so far:
> 
> * When passing URLs to already running Luakit: it results in empty page and I have to reload it to display it
> * Much higher CPU usage than on WebKit 2.24

I have precisely this 2 Issues ... had to downgrade to 2.24.

**Comment by gleachkr** [comment]
Similar CPU issues here, with the webkit process running at 40-60% on completely loaded pages that use negligible CPU in other browsers, fixed by downgrading webkit2gtk to 2.24.

**Comment by cyberpunkrocker-zero** [comment]
When using webkit2gtk-2.26.x + libsoup-2.68.x (on Arch Linux) I get completely blank screen on luakit, with only red "[notrust]" warning at the bottom; happened on every website I was trying to go. With combination webkit2gtk-2.24.x + libsoup-2.67.x luakit works as intended.
I found somewhere an old bug report describing a similar issue, libsoup being the culprit then. But that was years ago, and has been fixed since. I'm not sure which one is wrong here this time, webkit2gtk or libsoup.

**Comment by aidanholm** [comment]
That sounds like this bug; the root cause is outlined in my earlier comment. I'm currently experimenting with ways to fix this, but it's still pretty experimental (read: buggy).

**Comment by nst0022** [comment]
Same here on Debian after upgrade to
**libwebkit2gtk**-4.0-37:amd64  **2.26**.2-1~deb10+1
(1) page is empty until reloaded
(2) high CPU when window is open, low if iconified

**Comment by zappathustra** [comment]
Now webkit2gtk 2.24 seems to be broken in Arch (missing `libicuuc.so`, iirc), so you have to run/build with 2.26. And like @nst0022 pages won't load at first in an new tab. Next requests are ok, pages are loaded, but it's annoying when opening in an new tab (you have to reload once each time).

Tried switching to the official pacman binary but it's the same problem.

**Comment by nst0022** [comment]
Thanks for this hack, which works once, when luakit is started with a URL.

However, if I click a link on a page with the middle mouse button, which should open the link in a new tab, then I have to manually reload the tab for the content to be displayed.

Note, that I have re-applied the hack here locally in the version from the Debian repository, which is 
```
luakit 2.0.0
  built with webkit 2.22.6 
  (installed version: 2.26.2)
```
The version from GitHub might run better.

**Comment by Duchys** [comment]
Had the same problem when starting luakit with URL on Rasbian since upgrading webkit2gtk to version 2.26

Simply disabling adblock in rc.lua resolved this problem for me.

**Comment by nst0022** [comment]
@Duchys: Your are amazing. 
Indeed this works, even without the hack (and if you don't need ad-blocking, of course).

**Comment by nst0022** [comment]
@Duchys : You are amazing.
Indeed this works, even without the hack.

**Comment by zappathustra** [comment]
@nst0022 - The version in Debian is quite old, I can't really investigate. But anyway:
@Duchys - Indeed the culprit seems to be:

```
    webview.modify_load_block(view, "adblock", true)
```

on line 462 of `adblock.lua`. But then I don't know what `webview.modify_load_block()` does, so I can't help further (I'm not familiar with `webview.lua`). Quite possibly it simply loads whatever `adblock` is supposed to do, so the real problem is elsewhere in `adblock.lua`

**Comment by nst0022** [comment]
Not, that it gets lost, the high CPU usage problem still exists, with or without the disable-adblock-workaround.

**Comment by aidanholm** [comment]
@zappathustra see https://github.com/luakit/luakit/issues/804#issuecomment-533036061 for what `modify_load_block()` does. Basically, the problem is caused by the (now per-tab) process not being started, so adblock rules are never loaded, and the adblock module never finishes loading.
 
A partial fix is to add this to `init_funcs` in `lib/webview.lua`: 
```lua

    ensure_web_process_loaded = function(view)
        view:add_signal("switched-page", function (v)
            if not webview_state[v].process_loaded then
                if v.uri ~= nil then
                    v.uri = ""
                end
                webview_state[v].process_loaded = true
            end
        end)
    end,
```
This adds an extra `"about:blank"` history item to most tabs, so it's not ideal.

**Comment by aidanholm** [comment]
One possible alternative would be to navigate to `about:blank` in C, and hide the additional tab history item from lua.

**Comment by zappathustra** [comment]
@aidanholm I see, sorry I didn't read the thread properly!

**Comment by alex3kov** [comment]
FWIW - the high CPU issue is also present in Midori 8.0 but not in Epiphany 3.35.1 (using webkit2gtk 2.26.2).

**Comment by gleachkr** [comment]
Also (adding to @alex3kov), the CPU issue appears to not be present in surf 2.0.

**Comment by aidanholm** [comment]
@zappathustra No worries:)

CPU usage is likely an artefact of how luakit polls for changes to document height, in `extension_scroll.cc`. This seems to cause constant repaints starting from 2.26. That should be slightly more straight-forward to fix, by not using RAF, and adding a mechanism to restrict polling to currently visible webview widgets (which RAF does).

**Comment by msdemlei** [comment]
Here's a dumb question: Wouldn't setUsesSingleWebProcess(true) help?  I
freely admit I've not actually researched the situation but just looked
at a patch of the Debian webkit package, as in here:

https://git.launchpad.net/ubuntu/+source/webkit2gtk/tree/debian/patches/force-single-process.patch?h=debian/buster

**Comment by msdemlei** [comment]
> Here's a dumb question: Wouldn't setUsesSingleWebProcess(true) help? I
> freely admit I've not actually researched the situation but just looked
> at a patch of the Debian webkit package, as in here:
> 
> https://git.launchpad.net/ubuntu/+source/webkit2gtk/tree/debian/patches/force-single-process.patch?h=debian/buster

Replying to myself, that's clearly at least not nearly enough: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=945509

**Comment by hollmmax** [comment]
> 
> 
> @zappathustra No worries:)
> 
> CPU usage is likely an artefact of how luakit polls for changes to document height, in `extension_scroll.cc`. This seems to cause constant repaints starting from 2.26. That should be slightly more straight-forward to fix, by not using RAF, and adding a mechanism to restrict polling to currently visible webview widgets (which RAF does).

Any updates on this? If no, where should I look to do this? I haven't made any contributions to LuaKit yet and I haven't been using it for long either, so I don't know my way around the code base.

**Comment by sprnza** [comment]
Is hints color got black related issue?

**Comment by nst0022** [comment]
@HollMax: I guess with "extension_scroll.cc" @aidanholm meant "extension/scroll.c", where there is a function "check_for_document_resize" you could start with.

**Comment by hollmmax** [comment]
Basically the functions we use have been deprecated in 2.22, so they might have been broken in some way. Unfortunately there's no recommended replacement in WebkitGTK's documentation, so I can do a quick replacement and I don't have enough time to take a deeper look at what and why we're doing in extension/scroll.c and how to do it using JavascriptCore APIs.

**Comment by craftyguy** [comment]
> @zappathustra No worries:)
> 
> CPU usage is likely an artefact of how luakit polls for changes to document height, in `extension_scroll.cc`. This seems to cause constant repaints starting from 2.26. That should be slightly more straight-forward to fix, by not using RAF, and adding a mechanism to restrict polling to currently visible webview widgets (which RAF does).

Unsurprisingly, reverting f5f0239f19e3000de915067533e7d54e59e13c15 fixes the high CPU usage when displaying (any) content in a tab. This commit references [a webkit bugzilla to implemet ResizeObserver](https://bugs.webkit.org/show_bug.cgi?id=157743), looks like support was merged in April of this year but I'm not sure if it's in 2.26..

Anywys, reverting that and adding the hack @ #815 at least makes luakit usable again on webkit2gtk, even if just barely!

**Comment by nst0022** [comment]
I can confirm, that reverting f5f0239 solves the CPU usage issue on Debian 10 Buster.
Thanks for your efforts, @craftyguy .

**Comment by aidanholm** [comment]
Thanks for the info everyone; I've merged that PR (#804) and reverted that commit causing the high cpu usage.

I have a WIP fix in progress in the `wip-fix-804` branch that seems to be mostly working, but note that it's still a little bit unstable.

**Comment by bodqhrohro** [comment]
> Is hints color got black related issue?

@sprnza I got this even on 2.24 after several custom rebuilds and upgrades/downgrades with no apparent reason. Maybe my configuration got spoiled somewhere. Thought it's some layer compositing issue, but tried to override the hint styles with a userstyle recently, and voila! — it works.

```css
#luakit_select_overlay .hint_overlay {
    background-color: #ffff99 !important;
    border: 1px dotted #000 !important;
    opacity: 0.3 !important;
}

#luakit_select_overlay .hint_label {
    background-color: rgba(0, 0, 136, 0.4) !important;
    border: 1px dashed #000 !important;
    color: #fff !important;
    text-shadow: 1px 1px 2px black !important;
}

```

**Comment by ptaczek** [comment]
Hello everybody, 

Im running:
<pre><code>luakit 2.2
  built with webkit 2.28.4 (installed version: 2.28.4)</code></pre>
on Ubuntu 20.04

and I solved the luakit-not-loading-webpages issue by commenting out the adblock and adblock_chrome requires in rc.lua which brings back the unbearable lightness of browsing (thanks for the tips!)

and Im using AdGuard Home to block the ads globally. Or one can use Pi-hole to get the same ad-protection.

No CPU issues:
![gotop-luakit2 2](https://user-images.githubusercontent.com/1380733/91914817-44775e80-ecb9-11ea-8b64-12acb25ed10f.png)

#lovinluakitagain

**Comment by c0dev0id** [comment]
@sprnza  @bodqhrohro Thanks! I took the opacity for the link_overlay over into luakit, this should be enough to fix  the issue. Further customizations are on your own :) https://github.com/luakit/luakit/commit/16617a0a3beca75107834c81a857e1ed878fa845

@ptaczek If you have issues with adblock, please open a new issues.

I consider this issue to be solved with luakit 2.2 and therefore close it.
